### PR TITLE
separate llm and embedder

### PR DESCRIPTION
**Pull request by cyyeh**
With this PR, users can choose LLM and Embedding Model using different providers; for example. LLM using Ollama, Embedding Model using OpenAI.

updated corresponding documentation: https://github.com/Canner/docs.getwren.ai/pull/22

### CODE REVIEW

File: wren-ai-service/src/providers/llm/openai.py [hunk 1]
```diff
@@ -293,44 +129,34 @@ def _verify_api_key(api_key: str, api_base: str) -> None:
...
-        self._embedding_model_dim = embedding_model_dim
         self._generation_model = generation_model
 
     def get_generator(
         self,
-        model_kwargs: Optional[Dict[str, Any]] = GENERATION_MODEL_KWARGS,
+        model_kwargs: Dict[str, Any] = orjson.loads(
+            os.getenv("GENERATION_MODEL_KWARGS", "{}")
+        )
+        or GENERATION_MODEL_KWARGS,
```
Comment (paopa) [inline wren-ai-service/src/providers/llm/openai.py:146]: we assumed a default value when the env var is not exist thus I don't think the OR operator works here.

File: wren-ai-service/src/utils.py [hunk 2]
```diff
@@ -53,23 +53,26 @@ def load_env_vars() -> str:
...
+        os.getenv("EMBEDDER_PROVIDER", "openai_embedder")
+    )()
     document_store_provider = loader.get_provider(
         os.getenv("DOCUMENT_STORE_PROVIDER", "qdrant")
     )()
-    engine = loader.get_provider(os.getenv("ENGINE", "wren-ui"))()
+    engine = loader.get_provider(os.getenv("ENGINE", "wren_ui"))()
 
-    return llm_provider, document_store_provider, engine
+    return llm_provider, embedder_provider, document_store_provider, engine
```
Comment (paopa) [inline wren-ai-service/src/utils.py:75]: there are a few place I think we forgot to change, could you change them in the followup generation, generation, and sql correction pipeline as well? thanks!

File: wren-ai-service/src/providers/llm/ollama.py [hunk 3]
```diff
@@ -126,182 +119,29 @@ async def run(
...
         self._generation_model = generation_model
-        self._embedding_model = embedding_model
 
     def get_generator(
         self,
-        model_kwargs: Optional[Dict[str, Any]] = GENERATION_MODEL_KWARGS,
+        model_kwargs: Dict[str, Any] = orjson.loads(
+            os.getenv("GENERATION_MODEL_KWARGS", "{}")
+        )
+        or GENERATION_MODEL_KWARGS,
```
Comment (paopa) [inline wren-ai-service/src/providers/llm/ollama.py:138]: it is a gentle comment to mention here has the same situation that we mentioned in other comment! 😆

File: wren-ai-service/src/providers/llm/azure_openai.py [hunk 4]
```diff
@@ -125,228 +112,38 @@ async def run(
...
-        self._embedding_model = embedding_model
-        self._embedding_model_dim = embedding_model_dim
 
     def get_generator(
         self,
-        model_kwargs: Optional[Dict[str, Any]] = AZURE_GENERATION_MODEL_KWARGS,
+        model_kwargs: Dict[str, Any] = orjson.loads(
+            os.getenv("GENERATION_MODEL_KWARGS", "{}")
+        )
+        or GENERATION_MODEL_KWARGS,
```
Comment (paopa) [inline wren-ai-service/src/providers/llm/azure_openai.py:136]: it is a gentle comment to mention here has the same situation that we mentioned in other comment! 😆

**Review by paopa — COMMENTED** [review]
Reviewed part of the changes. and remaining code will been reviewed in tmr.

Could @onlyjackfrost also take a look at the section on the launcher and Docker? Thanks a lot.

File: wren-ai-service/src/providers/llm/openai.py (hunk 1 above)
Comment (paopa) [reply wren-ai-service/src/providers/llm/openai.py:146]: ignore the above comment, I am wrong, it works. 😅

```python
>>> {} or {'f':1}
{'f': 1}
```

**Review by paopa — APPROVED** [review]
overall LGTM. thanks for your hard work!

File: docker/docker-compose-dev.yaml [hunk 5]
```diff
@@ -44,10 +44,10 @@ services:
     environment:
       WREN_AI_SERVICE_PORT: ${WREN_AI_SERVICE_PORT}
       WREN_UI_ENDPOINT: ${WREN_UI_ENDPOINT}
-      GENERATION_MODEL: ${GENERATION_MODEL}
-      OPENAI_API_KEY: ${OPENAI_API_KEY}
-      AZURE_CHAT_KEY: ${AZURE_CHAT_KEY}
-      AZURE_EMBED_KEY: ${AZURE_EMBED_KEY}
+      LLM_OPENAI_API_KEY: ${LLM_OPENAI_API_KEY}
```
Comment (onlyjackfrost) [inline docker/docker-compose-dev.yaml:47]: @cyyeh  could you help me understand the differences between `LLM_OPENAI_API_KEY` and `EMBEDDER_OPENAI_API_KEY` and the usage of it?

Comment (cyyeh) [reply docker/docker-compose-dev.yaml:47]: @onlyjackfrost  After this PR, all of the configurations of LLM and EMBEDDER are separated. So users can flexibly choose what LLM and Embedder they need using different providers. For example. LLM using OpenAI, Embedder using Fireworks.ai(OpenAI api compatible).

**Review by onlyjackfrost — APPROVED** [review]
LGTM

File: docker/docker-compose-dev.yaml (hunk 5 above)
Comment (onlyjackfrost) [reply docker/docker-compose-dev.yaml:47]: Thanks for sharing

**Pull request merged**